src/adapter_gen.rs

```diff
@@ -1,40 +1,40 @@
 //! Generic Hardware Abstraction Layer, no_std-compatible.
 
-use crate::encoding::encode_rgb_slice;
+use crate::encoding::*;
 use alloc::boxed::Box;
 use alloc::string::String;
 
 /// SPI-device abstraction.
 pub trait HardwareDev {
-    fn write_all(&mut self, encoded_data: &[u8]) -> Result<(), String>;
+    type Error; 
+    fn write_all(&mut self, encoded_data: &[u8]) -> Result<(), Self::Error>;
 }
 
 pub trait WS28xxAdapter {
     /// Returns a reference to the hardware device.
     /// This function only needs to be implemented once in the generic adapter.
-    fn get_hw_dev(&mut self) -> &mut Box<dyn HardwareDev>;
+    fn get_hw_dev(&mut self) -> &mut Box<dyn HardwareDev<Error=String>>;
 
-    /// Encodes RGB values and write them via the hardware device to the LEDs. The length of the vector
+    /// Encodes LED values and write them via the hardware device to the LEDs. The length of the vector
     /// is the number of LEDs you want to write to. *Note* that if you have performance critical
     /// applications (like you need a signal on the LEDS on a given time) it's a better idea
-    /// to encode the data earlier by yourself using `crate::encoding`-module and calling
-    /// `WS28xxAdapter::write_encoded_rgb`. Otherwise and if your device is slow the encoding
-    /// could cost a few microseconds to milliseconds - depending on your amount of data and machine.
-    fn write_rgb(&mut self, rgb_data: &[(u8, u8, u8)]) -> Result<(), String> {
-        let encoded_data = encode_rgb_slice(rgb_data);
-        self.write_encoded_rgb(&encoded_data)
+    /// to encode the data earlier by yourself using [`crate::encoding`]-module 
+    /// and calling [`WS28xxAdapter::write_encoded_slice`].
+    fn write_slice(&mut self, data: &[u8]) -> Result<(), String> {
```

Comment:
You should probably go through this library and change out the `String`s for `&str`s or another error type, unless you want to be permanently straddled with the `alloc` crate (which ~~many~~ most embedded platforms don't support out of the box)